### PR TITLE
chore: Component.finance, Smoothy.finance, Saddle.finance, Curve open pools, Gas schedule [TKR-1]

### PR DESCRIPTION
**Pull request by romainbutteaud**
## Description

- Component.finance
- Smoothy.finance
- Curve Open Pools (STABLEx, alUSD) 
- Adjusting Gas Schedule

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Comment by romainbutteaud** [comment]
[Simbot](https://metabase.spaceship.0x.org/dashboard/139?run_id=romain_03_23_003&adjusted_amount_bought_win_tolerance__bps_=0.5)

Pairs tested: 
DAI/USDC, DAI/USDT, DAI/PAX, USDC/USDT, USDC/PAX, USDT/PAX, DAI/TUSD, USDC/TUSD, USDT/TUSD, SUSD/DAI, SUSD/USDC, SUSD/USDT, RENBTC/WBTC, SBTC/RENBTC, SBTC/WBTC, HBTC/WBTC, GUSD/DAI, GUSD/USDC, GUSD/USDT, HUSD/DAI, HUSD/USDC, HUSD/USDT, USDN/DAI, USDN/USDC, USDN/USDT, MUSD/DAI, MUSD/USDC, MUSD/USDT, DUSD/DAI, DUSD/USDC, DUSD/USDT, TBTC/RENBTC, TBTC/WBTC, TBTC/SBTC, PBTC/RENBTC, PBTC/WBTC, PBTC/SBTC, BBTC/RENBTC, BBTC/WBTC, BBTC/SBTC, OBTC/RENBTC, OBTC/WBTC, OBTC/SBTC, UST/DAI, UST/USDC, UST/USDT, EURS/SEUR, USDP/DAI, USDP/USDC, USDP/USDT, STABLEX/DAI, STABLEX/USDC, STABLEX/USDT, ALUSD/DAI, ALUSD/USDC, ALUSD/USDT, BUSD/USDC, BUSD/DAI, BUSD/USDT

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinSmoothy.sol
Comment (dekz) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinSmoothy.sol:68]: If this is the only difference (`swap` function) couldn't we re-use the `MixinCurve` 

```
            data.curveAddress.call(abi.encodeWithSelector(
                data.exchangeFunctionSelector,
                data.fromCoinIdx,
                data.toCoinIdx,
                // dx
                sellAmount,
                // min dy
                1
            ));
```

So we just need to get the 

```
function swap(
        uint256 bTokenIdxIn,
        uint256 bTokenIdxOut,
        uint256 bTokenInAmount,
        uint256 bTokenOutMin
    )
``` 

bytes4 and encode that as `exchangeFunctionSelector`. 

```
0x5673b02d  =>  swap(uint256,uint256,uint256,uint256)  
```

File: packages/asset-swapper/contracts/src/SmoothySampler.sol
Comment (dekz) [inline packages/asset-swapper/contracts/src/SmoothySampler.sol:95]: I believe `getSwapAmount` is a sell only function, as in it takes `fromAmount` and spits out `toAmount`. What we're trying to find in `samplerBuysFromSmoothy` is `toAmount` and spit out `fromAmount`, ie, I want 100 X, how much Y do I have to sell.

Some sources don't support this (like later Curves) so we approximate it. Check out  `function sampleBuysFromCurve` in CurveSampler.

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinSmoothy.sol
Comment (romainbutteaud) [reply contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinSmoothy.sol:68]: For some reason Curve uses `int128` for its coin index but Smoothy uses `int256`. So I can't reuse the `MixinCurve`, is that right?

File: packages/asset-swapper/contracts/src/SmoothySampler.sol
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/SmoothySampler.sol]: thanks @dekz, fixed it and tested `buys` on Simbot 👍

File: contracts/zero-ex/contracts/src/transformers/bridges/BridgeSource.sol
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/BridgeSource.sol:45]: might collide with #150 FYI

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinSmoothy.sol
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinSmoothy.sol:68]: You probably can.

File: packages/asset-swapper/src/utils/market_operation_utils/orders.ts
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/orders.ts:114]: ?
```suggestion
            return BridgeSource.Cpmponent;
```

**Comment by romainbutteaud** [comment]
updated after the BSC rebase, do not merge yet, still running simbots on all sources / pools.

**Comment by romainbutteaud** [comment]
Added a new Sampler for Smoothy.finance to only use what's available in their reserve.

**Review by dorothy-zbornak — APPROVED** [review]
some Qs but looking good m8

**Comment by dorothy-zbornak** [comment]
don't forget ye changelog

File: packages/asset-swapper/contracts/src/ShellSampler.sol
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/ShellSampler.sol:103]: That fixes the `buys` for Shell and the Shell forks. 
Right now since we have that 0 "slippage" on the `buys` for some routes, if we use the `Shell` buy function `viewTargetSwap`, it rounds up (tiny higher approximation than the settlement `targetSwap`), hence the 100% reverts on buys. 

```
IncompleteFillBuyQuoteError({
  buyToken: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
  boughtAmount: 1000667315,
  buyAmount: 1000667445
})

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts
Comment (romainbutteaud) [reply packages/asset-swapper/src/utils/market_operation_utils/constants.ts:282]: there's still lots of liquidity ($127m), but true that the gas cost is so high (compared to the 3crv pool) that that pool might never be used. Will remove it.

File: packages/asset-swapper/contracts/src/SmoothySampler.sol
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/SmoothySampler.sol]: updated

File: packages/asset-swapper/contracts/src/SmoothySampler.sol
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/SmoothySampler.sol]: updated

File: packages/asset-swapper/contracts/src/SmoothySampler.sol
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/SmoothySampler.sol]: Smoothy keeps some assets as y tokens (for example yUSDT). Sometimes there are no more USDT in the pool but only yUSDT. Problem was the yUSDT vault had some issue where the USDT was "locked" because it was borrowed and couldn't be withdrawn. The `getSwapAmount` quotes us assuming the USDT is withdrawable, so in the sampler I made sure we only look at that's in the contract (without checking yUSDT) so it can never revert. This is to be on the safe side. Also withdrawing from yUSDT would add a lot of gas overhead so not worth it.

Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/SmoothySampler.sol]: https://smoothy-finance.medium.com/smoothy-finance-withdrawal-incident-analysis-d9a85ab52983

Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/SmoothySampler.sol]: https://etherscan.io/address/0xe5859f4efc09027a9b718781dcb2c6910cac6e91

**Pull request merged**
